### DISCUSSION TITLE
DOM shuffled after hydration

### DISCUSSION THREAD

**Discussion by Xowap** [nuxt/nuxt#22019] [Questions]
Hi all,

I've got an extremely weird bug happening:

- In dev mode, all fine
- In prod mode
    - SSR DOM is fine
    - But after hydration the DOM gets shuffled

Basically I end up with things like:

```
<div class="foo"></div>
<div class="bar"></div>
```

Being turned into

```
<div class="foo"></div>
<div class="foo"></div>
```

I've managed to trim down the issue to [the following demo](https://github.com/Xowap/nuxt-hydration-bug). It looks to me that HTML comments are causing the issue, so I guess this could be related to some node-counting issue?

**Comment by TheAlexLichter** [comment]
That sounds like a hydration error I've described [in a blog post](https://blog.lichter.io/posts/vue-hydration-error/). Hard to pinpoint without dissecting the page step by step sadly 🙁

PS: The blog post was written for Nuxt 2 back then, so the hydration debugging in the console doesn't work 1:1 as explained but the idea is the same

**Reply by Xowap** [reply]
Thanks I'll have a look!

**Reply by Xowap** [reply]
Well I had a look, it's not really this. We have no actual hydration warning and as a matter of fact it works absolutely great in dev mode. It's just in prod mode that we're experiencing issues... It really looks like a bug inside of Vue/Nuxt, as the dev mode works well and shows no warnings.

Not sure if there are ways to debug this in the sense of "why the fuck does it decide to duplicate random nodes in the DOM"?

**Reply by TheAlexLichter** [reply]
> We have no actual hydration warning and as a matter of fact it works absolutely great in dev mode. It's just in prod mode that we're experiencing issues...

This is the point. It is common to work in dev but not in prod if it is a hydration error.

If you can share a reproduction, I am happy to have a look

**Reply by Xowap** [reply]
Honestly I've got a few 100ks LoC of Nuxt 2 under my belt and I've seen the things you describe in your post and what we got here really doesn't feel like that. Specifically, there is no error, it's just that the generated DOM doesn't make any sense (shuffled/duplicated items).

I understand that without a reproduction it's impossible to look at the issue, however I'm not sure how to isolate my case, which is kind of the direction of my question. It happens in a big, complex DOM and the most isolated case I've been able to produce is still fairly big.

I'll keep looking

**Comment by Xowap** [comment]
I've created a demo of the bug: https://github.com/Xowap/nuxt-hydration-bug

It looks like the issue is eventually caused by HTML comments ?

**Comment by Xowap** [comment]
After investigating a bit, it seems to be a [Vue bug](https://github.com/vuejs/core/issues/9185#issuecomment-1713717699).
